File: app/src/main/java/com/raival/fileexplorer/tab/file/task/ExtractTask.java
package com.raival.fileexplorer.tab.file.task;

import android.os.Handler;
import android.os.Looper;

import com.raival.fileexplorer.tab.file.model.Task;
import com.raival.fileexplorer.util.ZipUtils;

import java.io.File;
import java.util.ArrayList;

public class ExtractTask extends Task {
    private final ArrayList<File> filesToExtract;
    private File activeDirectory;

    public ExtractTask(ArrayList<File> filesToExtract) {
        this.filesToExtract = filesToExtract;
    }

    public ArrayList<File> getFilesToExtract() {
        return filesToExtract;
    }

    @Override
    public String getName() {
        return "Extract";
    }

    @Override
    public String getDetails() {
        StringBuilder sb = new StringBuilder();
        boolean first = true;

        for (File file : filesToExtract) {
            if (!first) {
                sb.append(", ");
            }
            sb.append(file.getName());
            first = false;
        }
        return sb.toString();
    }

    @Override
    public boolean isValid() {
        for (File file : filesToExtract) {
            if (!file.exists())
                return false;
        }
        return true;
    }

    @Override
    public void setActiveDirectory(File directory) {
        activeDirectory = directory;
    }

    @Override
    public void start(OnUpdateListener onUpdateListener, OnFinishListener onFinishListener) {
        new Thread(() -> {
            new Handler(Looper.getMainLooper()).post(() -> onUpdateListener.onUpdate("Extracting...."));
            try {
                ZipUtils.extract(filesToExtract, activeDirectory);
                new Handler(Looper.getMainLooper()).post(() -> onFinishListener.onFinish("Files have been extracted successfully"));
            } catch (Exception e) {
                e.printStackTrace();
                new Handler(Looper.getMainLooper()).post(() -> onFinishListener.onFinish(e.toString()));
            }
        }).start();
    }
}

File: app/src/main/java/com/raival/fileexplorer/tab/file/task/ExtractTask.kt
package com.raival.fileexplorer.tab.file.task

import android.os.Handler
import android.os.Looper
import com.raival.fileexplorer.tab.file.misc.extract
import com.raival.fileexplorer.tab.file.model.Task
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File

class ExtractTask(private val filesToExtract: ArrayList<File>) : Task() {
    private var activeDirectory: File? = null
    override val name: String
        get() = "Extract"
    override val details: String
        get() {
            val sb = StringBuilder()
            var first = true
            for (file in filesToExtract) {
                if (!first) {
                    sb.append(", ")
                }
                sb.append(file.name)
                first = false
            }
            return sb.toString()
        }
    override val isValid: Boolean
        get() {
            for (file in filesToExtract) {
                if (!file.exists()) return false
            }
            return true
        }

    override fun setActiveDirectory(directory: File) {
        activeDirectory = directory
    }

    override fun start(onUpdate: OnUpdateListener, onFinish: OnFinishListener) {
        CoroutineScope(Dispatchers.IO).launch {
            Handler(Looper.getMainLooper()).post { onUpdate.onUpdate("Extracting....") }
            try {
                extract(filesToExtract, activeDirectory!!)
                withContext(Dispatchers.Main) { onFinish.onFinish("Files have been extracted successfully") }
            } catch (e: Exception) {
                e.printStackTrace()
                withContext(Dispatchers.Main) { onFinish.onFinish(e.toString()) }
            }
        }.start()
    }
}